rust/lance/src/index/vector/ivf/io.rs

```diff
@@ -309,58 +320,190 @@ pub(super) async fn write_hnsw_index_partitions(
         )
         .await?;
 
-        let mut vector_batches = Vec::new();
+        let writer_options = FileWriterOptions::default();
+        let (part_file, aux_part_file) = build_part_path(part_id);
+        let part_writer = FileWriter::<ManifestDescribing>::try_new(
+            dataset.object_store(),
+            &part_file,
+            Schema::try_from(writer.schema())?,
+            &writer_options,
+        )
+        .await?;
 
-        let projection = dataset.schema().project(&[column])?;
-        for row_ids in &row_id_array {
-            let array = dataset
-                .take_rows(row_ids.as_primitive::<UInt64Type>().values(), &projection)
-                .await?
-                .column_by_name(column)
-                .expect("row id column not found")
-                .clone();
-            vector_batches.push(array);
-        }
+        let aux_part_writer = match aux_part_file {
+            Some(path) => Some(
+                FileWriter::<ManifestDescribing>::try_new(
+                    dataset.object_store(),
+                    &path,
+                    Schema::try_from(auxiliary_writer.as_ref().unwrap().schema())?,
+                    &writer_options,
+                )
+                .await?,
+            ),
+            None => None,
+        };
 
-        let params = shared_params.clone();
+        let dataset = dataset.clone();
+        let column = column.clone();
+        let hnsw_params = hnsw_params.clone();
         let pq = pq.clone();
-        let build_with_pq = auxiliary_writer.is_some();
-        let task = utils::tokio::spawn_cpu(move || {
-            build_hnsw_partition(
-                metric_type,
-                params,
-                row_id_array,
-                pq_array,
-                vector_batches,
-                build_with_pq,
-                pq,
-            )
-        });
-
-        task_queue.push_back(task);
+        tasks.push(build_hnsw_partition(
+            dataset,
+            column,
+            metric_type,
+            hnsw_params,
+            part_writer,
+            aux_part_writer,
+            pq,
+            row_id_array,
+            pq_array,
+        ));
     }
 
-    while let Some(task) = task_queue.pop_front() {
-        let (hnsw, pq_storage) = task.await?;
+    let parallel_limit = 16;
```

Comment:
How do we pick 16 here? Btw lets define this as `const`?

Reply:
it's up to how many cores and memory we have, something like `min(num_cpus, memory / partition_size)`